js/src/edit-free-campaign/hasUnsavedShippingRages.js

```diff
@@ -0,0 +1,43 @@
+/**
+ * External dependencies
+ */
+import { isEqual, differenceWith, at } from 'lodash';
+
+/**
+ * @typedef {import('.~/data/actions').ShippingRate} ShippingRate
+ */
+
+/**
+ * This function compares the key fields of two shipping rate arrays
+ * and returns whether there are any unsaved shipping changes.
+ *
+ * - Order:     The order of two arrays does not need to be the same.
+ * - ID:        The `id` is not compared.
+ *              On the UI, if a rate was deleted and then added,
+ *              then the `id` would not exist.
+ * - Threshold: If `options` or `options.free_shipping_threshold` does not exist,
+ *              `free_shipping_threshold` is treated as undefined.
+ *
+ * @param  {Array<ShippingRate>} rates The shipping rates to be compared.
+ * @param  {Array<ShippingRate>} savedRates The saved shipping rates received from API.
+ * @return {boolean} Whether there are any unsaved shipping changes.
+ */
+export default function hasUnsavedShippingRages( rates, savedRates ) {
+	if ( rates.length !== savedRates.length ) {
+		return true;
+	}
+
+	const paths = [
+		'country',
+		'method',
+		'currency',
+		'rate',
+		'options.free_shipping_threshold',
+	];
```

Comment:
❔ The approach here is to compare specific known paths, i.e. using an "inclusive" approach.

When I first thought about this issue, I was thinking to compare everything in the shipping rate object except `id`, i.e. an "exclusive" approach. 

Do you think if there is any pros and cons to the above? Just curious to hear your opinion. 🤔

Reply:
In the long term, the exclusive approach should be ideal for shipping data. Not only compatibility toward the future changes, but It's also easier to find out if there are new exclusive properties, like the case in this time, the need to make changes is due to the full exclusivity of the original `isEqual`.

I was a bit hesitant while developing because when the form handles `options.free_shipping_threshold`, the way to clear threshold is to set it to `undefined`, but the API is to unset the property. Two options came to mind at the moment:
1. Adjust the logic in forms - There were other related developments still in progress, in addition to further reducing the flexibility of form handling data.
2. Normalize shipping rates before comparing - But by doing so, it is still an inclusive approach.

So in the short term, I chose inclusive approach first, and we could change it to exclusive approach when someday it's applicable.